pkg/cluster/clustermsi.go

```diff
@@ -0,0 +1,172 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/msi-dataplane/pkg/dataplane"
+	"github.com/Azure/msi-dataplane/pkg/store"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/azuresdk/armmsi"
+)
+
+const (
+	msiCertValidityDays = 90
+)
+
+// ensureClusterMsiCertificate leverages the MSI dataplane module to create a new MSI
+// certificate (if needed) and store the certificate in the cluster MSI key vault. It
+// does not concern itself with whether an existing certificate is valid or not; that
+// can be left to the certificate refresher component.
+func (m *manager) ensureClusterMsiCertificate(ctx context.Context) error {
+	secretName, err := m.clusterMsiSecretName()
+	if err != nil {
+		return err
+	}
+
+	_, err = m.clusterMsiKeyVaultStore.GetCredentialsObject(ctx, secretName)
+	if err == nil {
+		return nil
+	} else if azcoreErr, ok := err.(*azcore.ResponseError); !ok || azcoreErr.StatusCode != 404 {
+		return err
+	}
+
+	clusterMsiResourceId, err := m.clusterMsiResourceId()
+	if err != nil {
+		return err
+	}
+
+	uaMsiRequest := dataplane.UserAssignedMSIRequest{
+		IdentityURL: m.doc.OpenShiftCluster.Identity.IdentityURL,
+		ResourceIDs: []string{clusterMsiResourceId.String()},
+		TenantID:    m.doc.OpenShiftCluster.Identity.TenantID,
+	}
+
+	msiCredObj, err := m.msiDataplane.GetUserAssignedIdentities(ctx, uaMsiRequest)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+	expirationDate := now.AddDate(0, 0, msiCertValidityDays)
```

Comment:
You can get this date from the credentials object - use `not_after` in nested credentials object

Reply:
Are you sure? How do we know that it will be 90 days (since this value was agreed upon during the design phase)? Also, I think the stub would return a nil `NotAfter`, so we'd have to add a conditional to set this "manually" when using the stub.

Reply:
What if NotAfter is earlier than 90days?
Can't this happen?

Reply:
1. MSI team has a doc on this: https://eng.ms/docs/products/arm/rbac/managed_identities/msionboardingcertificaterotation. 

`day 90: The certificate expires in Azure AD and can no longer be used to get tokens. This is the not_after field in the credential response.`

2. What if the MSI team makes a change and says cert expiration is now at 45 days? Given the above doc, it's better to rely on a value received from them than to hardcode something on our end.

3. Re stub -> we can just set `NotAfter` to some arbitrary date in the future for testing (ex: one minute from now).

Reply:
Oh, I see. Then I'll use the `NotAfter` returned by the MSI RP.

Reply:
Fixed!